include/rmm/mr/device/aligned_resource_adaptor.hpp

```diff
@@ -0,0 +1,219 @@
+   *
+   * @param upstream The resource used for allocating/deallocating device memory.
+   * @param allocation_alignment The size used for allocation alignment.
+   * @param alignment_threshold Only allocations with a size larger than or equal to this threshold
+   * are aligned.
+   */
+  explicit aligned_resource_adaptor(Upstream* upstream,
+                                    std::size_t allocation_alignment = default_allocation_alignment,
+                                    std::size_t alignment_threshold  = default_alignment_threshold)
+    : upstream_{upstream},
+      allocation_alignment_{allocation_alignment},
+      alignment_threshold_{alignment_threshold}
+  {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+    RMM_EXPECTS(allocation_alignment % 256 == 0, "Allocation alignment is not a multiple of 256.");
+    RMM_EXPECTS(alignment_threshold % 256 == 0, "Alignment threshold is not a multiple of 256.");
+  }
+
+  aligned_resource_adaptor()                                = delete;
+  ~aligned_resource_adaptor() override                      = default;
+  aligned_resource_adaptor(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor(aligned_resource_adaptor&&)      = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor&&) = delete;
+
+  /**
+   * @brief Get the upstream memory resource.
+   *
+   * @return Upstream* pointer to a memory resource object.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @copydoc rmm::mr::device_memory_resource::supports_streams()
+   */
+  [[nodiscard]] bool supports_streams() const noexcept override
+  {
+    return upstream_->supports_streams();
+  }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  [[nodiscard]] bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+ private:
+  static constexpr std::size_t default_allocation_alignment = 256;
+  static constexpr std::size_t default_alignment_threshold  = 0;
+  using lock_guard                                          = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream resource with the specified
+   * alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    if (allocation_alignment_ == default_allocation_alignment || bytes < alignment_threshold_) {
```

Comment:
Well, it will align it how it wants, which may not match the if statement here. In the admittedly unlikely event that CUDA started to return 16 byte aligned memory, this code would pass through to it if the user requested 256-byte alignment unless this was remembered and adjusted. At that point this would be incorrect.

This is certainly a pedantic complaint at best, but I have seen bugs like this in the past and I would like to avoid such things as they are difficult to track. If we can't find the default alignment from CUDA, I am ok with letting this go.

Reply:
Good point. We have 256 hard coded all over the place. I added a constant in `aligned.hpp` and replaced all the hard-coded values.